rust/worker/src/execution/operators/knn_projection.rs

```diff
@@ -0,0 +1,253 @@
+use chroma_blockstore::provider::BlockfileProvider;
+use chroma_error::ChromaError;
+use chroma_types::Segment;
+use thiserror::Error;
+use tonic::async_trait;
+use tracing::trace;
+
+use crate::execution::{
+    operator::{Operator, OperatorType},
+    operators::projection::ProjectionInput,
+};
+
+use super::{
+    fetch_log::FetchLogOutput,
+    knn::RecordDistance,
+    projection::{ProjectionError, ProjectionOperator, ProjectionRecord},
+};
+
+/// The `KnnProjectionOperator` retrieves record content by offset ids
+/// It is based on `ProjectionOperator`, and it attaches the distance
+/// of the records to the target embedding to the record content
+///
+/// # Parameters
+/// - `projection`: The parameters of the `ProjectionOperator`
+/// - `distance`: Whether to attach distance information
+///
+/// # Inputs
+/// - `logs`: The latest logs of the collection
+/// - `blockfile_provider`: The blockfile provider
+/// - `record_segment`: The record segment information
+/// - `record_distances`: The offset ids of the record to retrieve for,
+///   along with their distances to the target embedding
+///
+/// # Outputs
+/// - `records`: The retrieved records in the same order as `record_distances`
+///
+/// # Usage
+/// It can be used to retrieve record contents as user requested
+/// It should be run as the last step of an orchestrator
+#[derive(Clone, Debug)]
+pub struct KnnProjectionOperator {
+    pub projection: ProjectionOperator,
+    pub distance: bool,
+}
+
+#[derive(Clone, Debug)]
+pub struct KnnProjectionInput {
+    pub logs: FetchLogOutput,
+    pub blockfile_provider: BlockfileProvider,
+    pub record_segment: Segment,
+    pub record_distances: Vec<RecordDistance>,
+}
+
+#[derive(Clone, Debug)]
+pub struct KnnProjectionRecord {
+    pub record: ProjectionRecord,
+    pub distance: Option<f32>,
+}
+
+#[derive(Clone, Debug)]
+pub struct KnnProjectionOutput {
+    pub records: Vec<KnnProjectionRecord>,
+}
+
+#[derive(Error, Debug)]
+pub enum KnnProjectionError {
+    #[error("Error running projection operator: {0}")]
+    Projection(#[from] ProjectionError),
+}
+
+impl ChromaError for KnnProjectionError {
+    fn code(&self) -> chroma_error::ErrorCodes {
+        match self {
+            KnnProjectionError::Projection(e) => e.code(),
+        }
+    }
+}
+
+#[async_trait]
+impl Operator<KnnProjectionInput, KnnProjectionOutput> for KnnProjectionOperator {
+    type Error = KnnProjectionError;
+
+    fn get_type(&self) -> OperatorType {
+        OperatorType::IO
```

Comment:
i think this should not be IO operator - discussed offline

Reply:
https://github.com/chroma-core/chroma/pull/3083

Reply:
Also changed this operator to `Other` type